level00/PropertyTests.hs

```diff
@@ -0,0 +1,153 @@
+{-# OPTIONS_GHC -Wno-unused-binds -Wno-unused-imports #-}
+{-# LANGUAGE RankNTypes #-}
+module PropertyTests (propertyTests) where
+
+import           Control.Applicative (liftA2)
+import           Control.Lens        (Prism', matching, preview, review)
+import           Control.Lens.TH     (makePrisms)
+
+import           Data.Function       (on)
+import           Data.List           (filter, insertBy, nubBy)
+
+import           Test.Tasty          (TestTree, testGroup)
+import           Test.Tasty.Hedgehog (testProperty)
+
+import           Hedgehog
+import qualified Hedgehog.Gen        as Gen
+import qualified Hedgehog.Range      as Range
+
+import           MyBTree
+
+-- [OPTIONAL]
+import           LawPropertiesBonus  (myBTreePrismLaws)
+
+----------------------------------------------------------------------------------------------------
+addTen :: Int -> Int
+addTen = (+10)
+
+-- A simple property for the 'addTen' function, ensure that:
+--
+-- \/ (x : Int) -> addTen x === (x + 10)
+prop_addTen :: Property
+prop_addTen = error "prop_addTen not implemented"
+
+----------------------------------------------------------------------------------------------------
+
+-- Property based testing is useful for debugging. In this exercise we've given
+-- you a broken function. Using the output of the test, fix up the function. If
+-- you're keen you can expand the properties that it must satisfy to make the
+-- test, and commensurately the function itself, more robust.
+--
+badReverse :: [a] -> [a]
+badReverse []     = []
+badReverse (x:xs) = x : reverse xs
+
+prop_badReverse :: Property
+prop_badReverse = property $ do
+  xs <- forAll (Gen.list (Range.linear 0 1000) Gen.bool)
+  badReverse (badReverse xs) === xs
+  -- [BONUS]: Are there other properties for reversing a list that could make this test more robust?
+
+----------------------------------------------------------------------------------------------------
+
+-- In this set of exercises we will build properties to ensure that our
+-- functions for a given data structure do what we expect.
```

Comment:
Cite source?

Reply:
Eeep, source added there, and to btree exercises & data structure.